server/controllers/doppler.controller.js

```diff
@@ -0,0 +1,27 @@
+class DopplerController {
+    constructor(redisClientFactory) {
+      this.redisClientFactory = redisClientFactory;
+    }
+
+    async getShops(request, response) {
+        const {
+            session: { dopplerApiKey }
+        } = request;
+
+        const redis = this.redisClientFactory.createClient();
+        const shops = (await redis.getShopsAsync(dopplerApiKey, false))
+        .map(async shopName => {
+            let shop = await redis.getShopAsync(shopName);
+            return {
+                shop: shopName,
+                accessToken: shop.accessToken,
+                connectedOn: shop.connectedOn
+            };
+        });

```

Comment:
How does this work? it shouts all request in parallel and then waits for all of them? or it shouts the requests one by one (waiting for each one)?

Is it a good practice combining _map_ and _async_ in this way?

Reply:
>How does this work? 
The idea is to create an array of promises and then fire them all in parallel with `Promise.all()`. As far as I understand the `async` word converts the function into a Promise. While debugging the value is something like `[Promise{}, Promise, [], ...]`

> Is it a good practice combining map and async in this way?
I honestly don't know, it's just a practice. Otherwise I had to iterate over the array, using await for each function execution, without taking advantage of `Promise.all()`

Do you have some alternative in mind?

Reply:
I have not enough experience on this kind of programming in JavaScript (and honestly neither in C#). In general, I prefer avoiding parallel requests if they are not really necessary, mainly when the number of simultaneous requests is undeterminated.

When the number of simultaneous requests is undeterminated, I think that if the background infrastructure is not prepared, it could be not scalable and there should be to take more things into account. For example, it is easy having more issues with timeouts.

I would use consecutive requests in place of parallel ones to avoid unexpected (for me) issues, but as I said I have not enough experience and maybe there are already know good practices that we could take advantage from.